src/core/kernel/fcram.cpp

```diff
@@ -0,0 +1,107 @@
+#include "fcram.hpp"
+#include "memory.hpp"
+
+void KFcram::Region::reset(u32 start, size_t size) {
+	this->start = start;
+	pages = size >> 12;
+	freePages = pages;
+
+	Block initialBlock(pages, 0);
+	blocks.clear();
+	blocks.push_back(initialBlock);
+}
+
+void KFcram::Region::alloc(std::list<FcramBlock>& out, s32 allocPages, bool linear) {
+	for (auto it = blocks.begin(); it != blocks.end(); it++) {
+		if (it->used) continue;
+
+		// On linear allocations, only a single contiguous block may be used
+		if (it->pages < allocPages && linear) continue;
+
+		// If the current block is bigger than the allocation, split it
+		if (it->pages > allocPages) {
+			Block newBlock(it->pages - allocPages, it->pageOffs + allocPages);
+			it->pages = allocPages;
+			blocks.insert(it, newBlock);
+		}
+
+		// Mark the block as allocated and add it to the output list
+		it->used = true;
+		allocPages -= it->pages;
+		freePages -= it->pages;
+
+		u32 paddr = start + (it->pageOffs << 12);
+		FcramBlock outBlock(paddr, it->pages);
+		out.push_back(outBlock);
+
+		if (allocPages < 1) return;
+	}
+
+	// Official kernel panics here
+	Helpers::panic("Failed to allocate FCRAM, not enough guest memory");
+}
+
+u32 KFcram::Region::getUsedCount() {
+	return pages - freePages;
+}
+
+u32 KFcram::Region::getFreeCount() {
+	return freePages;
+}
+
+KFcram::KFcram(Memory& mem) : mem(mem) {}
+
+void KFcram::reset(size_t ramSize, size_t appSize, size_t sysSize, size_t baseSize) {
+	fcram = mem.getFCRAM();
+	refs = std::unique_ptr<u32>(new u32[ramSize >> 12]);
+	memset(refs.get(), 0, (ramSize >> 12) * sizeof(u32));
```

Comment:
std::memset

Reply:
I think this file can also use a quick clang-format, seems like headers at the top at least are unformatted